src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -14,592 +14,549 @@ namespace Microsoft.CodeAnalysis.CSharp
-                                       locals.Remove(constructorLocal.LocalSymbol) &&
-                                       (object)objectCreationConstructor.OriginalDefinition == _compilation.GetSpecialTypeMember(SpecialMember.System_ReadOnlySpan_T__ctor_Reference) &&
-                                       objectCreationConstructor.ContainingType.IsReadOnlySpanChar():
-                                    var wrappedExpr = ConvertConcatExprToString(assignment.Right);
-                                    unwrappedArgsBuilder.Add(wrappedExpr);
-                                    break;
-                                default:
-                                    locals.Free();
-                                    unwrappedArgsBuilder.Free();
-                                    arguments = default;
-                                    return false;
-                            }
-                        }
-
-                        if (locals.Count > 0)
-                        {
-                            // Not all locals are part of a known shape
-                            locals.Free();
-                            unwrappedArgsBuilder.Free();
-                            arguments = default;
-                            return false;
-                        }
+                        binaryOperator = left;
+                    }
+                    else
+                    {
+                        self.AddConcatArgument(binaryOperator.Left, argumentAlreadyVisited: false, arguments, ref concatMethods);
+                        break;
+                    }
+                }
+                while (true);
 
-                        locals.Free();
-                        arguments = unwrappedArgsBuilder.ToImmutableAndFree();
+                static bool shouldRecurse(BoundExpression expr, [NotNullWhen(true)] out BoundBinaryOperator? binaryOperator)
+                {
+                    binaryOperator = expr as BoundBinaryOperator;
+                    if (IsBinaryStringConcatenation(binaryOperator) && binaryOperator.InterpolatedStringHandlerData is null)
+                    {
                         return true;
                     }
-
-                    break;
+                    else
+                    {
+                        binaryOperator = null;
+                        return false;
+                    }
+                }
             }
-
-            arguments = default;
-            return false;
         }
 
-        /// <summary>
-        /// folds two concat operands into one expression if possible
-        /// otherwise returns null
-        /// </summary>
-        private BoundExpression? TryFoldTwoConcatOperands(BoundExpression loweredLeft, BoundExpression loweredRight)
+        private void AddConcatArgument(BoundExpression argument, bool argumentAlreadyVisited, ArrayBuilder<BoundExpression> finalArguments, ref WellKnownConcatRelatedMethods wellKnownConcatOptimizationMethods)
```

Comment:
>AddConcatArgument [](http://example.com/codeflow?start=21&length=17)

The name of this method feels too "innocent". 
- It doesn't reflect the fact that it might decompose the passed in ```argument``` into multiple entries in ```finalArguments```
- It doesn't  reflect the fact that it might add nothing to ```finalArguments```
- It doesn't reflect the fact that existing entries in ```finalArguments``` are expected to be in reversed order
- It doesn't reflect the fact that we expect the method to be called for concatenation arguments in reversed order. I.e. the ```argument``` must immediately precede arguments that are already in the ```finalArguments```.

It would be good to come up with a more descriptive name and add a doc comment about the contract, etc.